File: Metro/src/main/java/org/cubexmc/metro/util/LocationUtil.java
package org.cubexmc.metro.util;

import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.block.Block;

import java.util.EnumSet;
import java.util.Set;

/**
 * 位置工具类，用于位置数据的序列化和反序列化
 */
public class LocationUtil {
    
    /**
     * 所有铁轨类型的Material集合，用于快速检查
     */
    private static final Set<Material> RAIL_MATERIALS = EnumSet.of(
        Material.RAIL,
        Material.POWERED_RAIL,
        Material.DETECTOR_RAIL,
        Material.ACTIVATOR_RAIL
    );
    
    /**
     * 将Location对象转换为字符串格式
     * 格式为: "世界名,x,y,z"
     */
    public static String locationToString(Location location) {
        if (location == null || location.getWorld() == null) {
            return null;
        }
        
        return String.format("%s,%d,%d,%d", 
                location.getWorld().getName(), 
                location.getBlockX(), 
                location.getBlockY(), 
                location.getBlockZ());
    }
    
    /**
     * 检查给定位置是否为铁轨方块
     */
    public static boolean isRail(Location location) {
        if (location == null) {
            return false;
        }
        
        Block block = location.getBlock();
        return RAIL_MATERIALS.contains(block.getType());
    }
    
    /**
     * 检查给定位置是否在铁轨上（包括当前位置和下方一格）
     */
    public static boolean isOnRail(Location location) {
        if (location == null) {
            return false;
        }
        
        // 检查当前位置
        if (isRail(location)) {
            return true;
        }
        
        // 检查下方一格
        Location belowLocation = location.clone().subtract(0, 1, 0);
        return isRail(belowLocation);
    }
    
    /**
     * 获取两个位置之间的方向向量
     */
    public static org.bukkit.util.Vector getDirectionVector(Location from, Location to) {
        return to.toVector().subtract(from.toVector()).normalize();
    }
} 

File: Metro/src/main/java/org/cubexmc/metro/util/LocationUtil.kt
package org.cubexmc.metro.util

import org.bukkit.Location
import org.bukkit.Material
import org.bukkit.util.Vector
import java.util.EnumSet

/**
 * 位置工具类，用于位置数据的序列化和反序列化
 */
object LocationUtil {
    private val RAIL_MATERIALS: Set<Material> = EnumSet.of(
        Material.RAIL,
        Material.POWERED_RAIL,
        Material.DETECTOR_RAIL,
        Material.ACTIVATOR_RAIL,
    )

    @JvmStatic
    fun locationToString(location: Location?): String? {
        val world = location?.world ?: return null
        return String.format("%s,%d,%d,%d", world.name, location.blockX, location.blockY, location.blockZ)
    }

    @JvmStatic
    fun isRail(location: Location?): Boolean {
        if (location == null) {
            return false
        }
        return RAIL_MATERIALS.contains(location.block.type)
    }

    @JvmStatic
    fun isOnRail(location: Location?): Boolean {
        if (location == null) {
            return false
        }
        if (isRail(location)) {
            return true
        }
        val belowLocation = location.clone().subtract(0.0, 1.0, 0.0)
        return isRail(belowLocation)
    }

    @JvmStatic
    fun getDirectionVector(from: Location, to: Location): Vector = to.toVector().subtract(from.toVector()).normalize()
}
